### PR TITLE
types: returns data too long error

### PR DESCRIPTION
**Pull request by coocood**
returns data too long error when inserting a string longer than field type length.

Fixes https://github.com/pingcap/tidb/issues/1832

### CODE REVIEW

**Comment by coocood** [comment]
@zimulala @tiancaiamao PTAL

File: util/types/datum.go [hunk 1]
```diff
@@ -762,12 +762,16 @@ func (d *Datum) convertToString(target *FieldType) (Datum, error) {
...
+	if target.Flen > 0 && target.Flen < len(s) {
+		err = ErrDataTooLong
+	}
 	s = truncateStr(s, target.Flen)
 	ret.SetString(s)
 	if target.Charset == charset.CharsetBin {
 		ret.k = KindBytes
 	}
-	return ret, nil
+	return ret, err
```
Comment (zimulala) [inline util/types/datum.go]: Add trace.

File: util/types/datum.go [hunk 2]
```diff
@@ -762,12 +762,16 @@ func (d *Datum) convertToString(target *FieldType) (Datum, error) {
 		return invalidConv(d, target.Tp)
 	}
 	// TODO: consider target.Charset/Collate
+	var err error
+	if target.Flen > 0 && target.Flen < len(s) {
+		err = ErrDataTooLong
```
Comment (shenli) [inline util/types/datum.go]: Use ErrDataTooLong.Gen() to log target.Flen and len(s) ?

Comment (coocood) [inline util/types/datum.go]: MySQL error only has column name and row number, which we don't have in this function.

**Comment by shenli** [comment]
LGTM
@zimulala PTAL

**Comment by zimulala** [comment]
LGTM

**Pull request merged**